State is shown off even when on

Om home app State is shown off even when the camera is on. Switch will work fine on setting on and off but next time home app is opened the status reads again "OFF"

@williamli Hello, 

I tried to install your fork using 
pi@raspberrypi:~ $ sudo npm install -g git+https://github.com/williamli/homebridge-mijia-camera.git

I was unable to install; getting errors such as:

npm ERR! code 128
npm ERR! Command failed: /usr/bin/git submodule update -q --init --recursive
npm ERR! fatal: Could not change back to '/root/.npm/_cacache/tmp/git-clone-2633048f': Permission denied

Also attempted by using the version “sudo npm install -g homebridge-mijia-camera@0.0.2”
In this case i got an error saying no matching version found

And username; “sudo npm install -g williamli/homebridge-mijia-camera”
Getting the same error  about permission denied.

May be it was too soon and I needed the fork code to be validated?
Regards,

--- a/devices/MijiaCamera.js
+++ b/devices/MijiaCamera.js
@@ -13,30 +13,38 @@ class MijiaCamera {
   async connect () {
     const { ip: address, token } = this
 
-    const device = await miio.device({ address, token }).catch(err => {
-      throw (new Error('Error creating device', err))
-    })
+    this.log(`connecting to camera at ${address}...`)
 
-    this.device = device
+    return miio.device({ address, token });
+
+    // this.device = device
   }
 
   async getPowerState () {
-    if (!this.device) {
-      this.log.error('Camera not available')
-      throw (new Error('Camera not available'))
-    }
 
-    return this.device.call('get_prop', ['power'])
+    try {
+      const device = await this.connect()
+      return device.call('get_prop', ['power'])
+    } catch (err) {
+      this.log.error('Camera not available')
+      throw (new Error('Camera not available', err))
+    }
+    
   }
 
   async setPowerState (state) {
-    if (!this.device) {
+
+    try {
+      const device = await this.connect()
+      device.call('set_power', [state ? 'on' : 'off'])
+    } catch (err) {
       this.log.error('Camera not available')
-      throw (new Error('Camera not available'))
+      throw (new Error('Camera not available', err))
     }
 
-    const status = state ? 'on' : 'off'
-    this.device.call('set_power', [status])
+
+    
+    
   }
 }
 

--- a/devices/MijiaCamera.js
+++ b/devices/MijiaCamera.js
@@ -22,6 +22,7 @@ class MijiaCamera {
 
   async getPowerState () {
     if (!this.device) {
+      this.log.error('Camera not available')
       throw (new Error('Camera not available'))
     }
 
@@ -30,6 +31,7 @@ class MijiaCamera {
 
   async setPowerState (state) {
     if (!this.device) {
+      this.log.error('Camera not available')
       throw (new Error('Camera not available'))
     }
 

--- a/devices/MijiaCamera.js
+++ b/devices/MijiaCamera.js
@@ -25,9 +25,7 @@ class MijiaCamera {
       throw (new Error('Camera not available'))
     }
 
-    const stats = await this.device.call('get_prop', ['power'])
-    this.log(`current power state: ${stats[0]}`)
-    return stats[0] === 'on'
+    return this.device.call('get_prop', ['power'])
   }
 
   async setPowerState (state) {

--- a/devices/MijiaCamera.js
+++ b/devices/MijiaCamera.js
@@ -1,12 +1,12 @@
 const miio = require('miio')
 
 class MijiaCamera {
-  constructor (config) {
+  constructor (config, log) {
     const { name, ip, token } = config
     if (!ip || !token) {
       throw new Error('The camera IP address and token are required')
     }
-
+    this.log = log;
     Object.assign(this, { name, ip, token })
   }
 
@@ -26,6 +26,7 @@ class MijiaCamera {
     }
 
     const stats = this.device.call('get_prop', ['power'])
+    this.log(`current power state: ${stats[0]}`)
     return stats[0] === 'on'
   }
 

--- a/devices/MijiaCamera.js
+++ b/devices/MijiaCamera.js
@@ -25,7 +25,7 @@ class MijiaCamera {
       throw (new Error('Camera not available'))
     }
 
-    const stats = this.device.call('get_prop', ['power'])
+    const stats = await this.device.call('get_prop', ['power'])
     this.log(`current power state: ${stats[0]}`)
     return stats[0] === 'on'
   }

--- a/index.js
+++ b/index.js
@@ -16,12 +16,6 @@ class MijiaCameraAccessory {
     this.serviceInfo = this.createServiceInfo()
     this.cameraPowerToggleService = this.createToggleService()
 
-    this.device.connect().then(() => {
-      this.log(`Successfully connected to camera ${config.ip}`)
-    }).catch(error => {
-      this.log('Error connecting to camera')
-      this.log(error)
-    })
   }
 
   createServiceInfo () {
@@ -48,6 +42,7 @@ class MijiaCameraAccessory {
     }
 
     const setCameraStatus = function (state, callback) {
+      this.log(`set power state to ${state}`)
       this.device.setPowerState(state)
         .then(() => {
           callback(null)

--- a/index.js
+++ b/index.js
@@ -40,7 +40,9 @@ class MijiaCameraAccessory {
     const getCameraStatus = function (callback) {
       this.device.getPowerState()
         .then(status => {
-          callback(null, status)
+          const power = status[0];
+          this.log(`current power state: ${power}`)
+          callback(null, power === 'on')
         })
         .catch(callback)
     }

--- a/index.js
+++ b/index.js
@@ -12,7 +12,7 @@ class MijiaCameraAccessory {
   constructor (log, config) {
     this.log = log
 
-    this.device = new MijiaCamera(config)
+    this.device = new MijiaCamera(config, this.log)
     this.serviceInfo = this.createServiceInfo()
     this.cameraPowerToggleService = this.createToggleService()
 

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "homebridge-mijia-camera",
-  "version": "0.0.1",
+  "version": "0.0.2",
   "description": "Xiaomi Mijia camera plugin for Homebridge",
   "license": "MIT",
   "keywords": [

